bundles/org.openhab.binding.vallox/src/main/java/org/openhab/binding/vallox/internal/se/connection/ValloxBaseConnector.java

```diff
@@ -0,0 +1,264 @@
+/**
+ * Copyright (c) 2010-2020 Contributors to the openHAB project
+ *
+ * See the NOTICE file(s) distributed with this work for additional
+ * information.
+ *
+ * This program and the accompanying materials are made available under the
+ * terms of the Eclipse Public License 2.0 which is available at
+ * http://www.eclipse.org/legal/epl-2.0
+ *
+ * SPDX-License-Identifier: EPL-2.0
+ */
+package org.openhab.binding.vallox.internal.se.connection;
+
+import java.io.IOException;
+import java.io.InputStream;
+import java.io.OutputStream;
+import java.util.ArrayList;
+import java.util.LinkedList;
+import java.util.List;
+import java.util.concurrent.ArrayBlockingQueue;
+import java.util.concurrent.ScheduledExecutorService;
+import java.util.concurrent.ScheduledFuture;
+import java.util.concurrent.TimeUnit;
+
+import org.eclipse.jdt.annotation.NonNullByDefault;
+import org.eclipse.jdt.annotation.Nullable;
+import org.eclipse.smarthome.io.transport.serial.SerialPortManager;
+import org.openhab.binding.vallox.internal.se.ValloxSEConstants;
+import org.openhab.binding.vallox.internal.se.telegram.SendQueueItem;
+import org.openhab.binding.vallox.internal.se.telegram.Telegram;
+import org.openhab.binding.vallox.internal.se.telegram.Telegram.TelegramState;
+import org.openhab.binding.vallox.internal.se.telegram.TelegramFactory;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * This interface defines methods to receive data from vallox.
+ *
+ * @author Miika Jukka - Initial contribution
+ */
+@NonNullByDefault
+public abstract class ValloxBaseConnector implements ValloxConnector {
+
+    private final Logger logger = LoggerFactory.getLogger(ValloxBaseConnector.class);
+
+    private final List<ValloxEventListener> listeners = new ArrayList<>();
+    private final LinkedList<SendQueueItem> sendQueue = new LinkedList<>();
+    protected final ArrayBlockingQueue<Byte> buffer = new ArrayBlockingQueue<>(1024);
```

Comment:
Actually I have not. What is there to gain?

Reply:
I only suggested it because of how you are using the buffer. Using an ArrayBlockingQueue as a byte buffer seems a bit clunky so using piped streams seems much more appropriate for the task. There aren't any huge advantages other than perhaps being more performant.

One thing I would suggest though is to not do message processing on the same thread that is reading from the stream. You don't know what kind of buffering the underlying input stream has so you shouldn't keep it waiting otherwise you risk the low level library buffers getting filled and dropping information or even killing your inputstream.

Reply:
I'm not fully convinced yet and correct me if I'm mistaken. Create dedicated threads for reading and processing the telegrams. Then pipe them together using `PipedInputStream / PipedOutputStream` and let the `InputStream#read()` methods of both block until there's something to read. And his only applies to `ValloxIpConnector` because `SerialPort` implementation is event based?

Reply:
You can keep as is if you want. You already have something working anyway. I just though I would mention my idea since it came to mind.

But yea, it would pretty much only apply `ValloxIpConnector` anyway.

Reply:
Ideas are always appreciated 👍